docs/src/apps/programming-faq.md

```diff
@@ -0,0 +1,70 @@
+---
+title: "Programming FAQ"
+---
+
+When writing or interacting with Solana programs, there are common questions or challenges that often come up.  Below are resources to help answer these questions.  If not addressed here, Solana [#developers Discord channel](https://discord.gg/RxeGBH) is a great resource.
+
+
+## InvalidInstructionData
+
+This program error can occur while trying to deserialize the instruction, check
+that the structure passed in matches exactly the instruction.  There may be some
+padding between fields.  Try packing and unpacking the instruction type `T` to
+determine the exact encoding the program expects
+
+```rust
+fn unpack<T>(input: &[u8]) -> &T;
+fn pack<T>(input: &T) -> &[u8];
```

Comment:
Is there a reason for not providing the simple implementation of these functions?

Reply:
`Pack` is now a trait in the SDK, each program may implement it differently:
https://github.com/solana-labs/solana/blob/master/sdk/src/program_pack.rs